compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs

```diff
@@ -386,6 +388,30 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
         Ok((has_changed, certainty, nested_goals))
     }
 
+    fn instantiate_response_discarding_overflow(
+        &mut self,
+        param_env: ty::ParamEnv<'tcx>,
+        source: GoalSource,
+        original_values: Vec<ty::GenericArg<'tcx>>,
+        response: CanonicalResponse<'tcx>,
+    ) -> Result<(Certainty, bool, Vec<Goal<'tcx, ty::Predicate<'tcx>>>), NoSolution> {
+        let keep_overflow_constraints = || {
```

Comment:
needs comment:

doing this only for normalize-to goals as that's the only thing needed for backcompat

doesn't require us to track the origin of a root obligation. If we were to also do this for trait goals, we'd have to worry about

```rust
trait Foo {}
trait Bar {}

trait FooBar {}
impl<T: Foo + Bar> FooBar for T {}

// These two should behave the same, rn we can drop constraints for both,
// but if we don't drop `Misc` goals we would only drop the constraints for
// `FooBar`
fn func1<T: Foo + Bar>() {}
fn func2<T: FooBaz>() {}
```

Reply:
pls add comment